### ISSUE TITLE
`weisfeiler_lehman_graph_hash` does not support `edge_attr` for `MultiDiGraph`

### ISSUE THREAD

**Issue by PhilippvK** [networkx/networkx#7593]
### Current Behavior

Using `nx.weisfeiler_lehman_graph_hash(G1, edge_attr="label")` leads to a `KeyError: 'label'` if `isinstance(G1, nx.MultiDiGraph)` .

<details>
  <summary>Full stack trace of exception</summary>

```
/usr/local/lib/python3.10/dist-packages/networkx/utils/decorators.py in argmap_weisfeiler_lehman_graph_hash_1(G, edge_attr, node_attr, iterations, digest_size, backend, **backend_kwargs)
      1 import bz2
      2 import collections
----> 3 import gzip
      4 import inspect
      5 import itertools

/usr/local/lib/python3.10/dist-packages/networkx/utils/backends.py in __call__(self, backend, *args, **kwargs)
    631         if not backends:
    632             # Fast path if no backends are installed
--> 633             return self.orig_func(*args, **kwargs)
    634 
    635         # Use `backend_name` in this function instead of `backend`

/usr/local/lib/python3.10/dist-packages/networkx/algorithms/graph_hashing.py in weisfeiler_lehman_graph_hash(G, edge_attr, node_attr, iterations, digest_size)
    152     subgraph_hash_counts = []
    153     for _ in range(iterations):
--> 154         node_labels = weisfeiler_lehman_step(G, node_labels, edge_attr=edge_attr)
    155         counter = Counter(node_labels.values())
    156         # sort the counter, extend total counts

/usr/local/lib/python3.10/dist-packages/networkx/algorithms/graph_hashing.py in weisfeiler_lehman_step(G, labels, edge_attr)
    143         new_labels = {}
    144         for node in G.nodes():
--> 145             label = _neighborhood_aggregate(G, node, labels, edge_attr=edge_attr)
    146             new_labels[node] = _hash_label(label, digest_size)
    147         return new_labels

/usr/local/lib/python3.10/dist-packages/networkx/algorithms/graph_hashing.py in _neighborhood_aggregate(G, node, node_labels, edge_attr)
     33     label_list = []
     34     for nbr in G.neighbors(node):
---> 35         prefix = "" if edge_attr is None else str(G[node][nbr][edge_attr])
     36         label_list.append(prefix + node_labels[nbr])
     37     return node_labels[node] + "".join(sorted(label_list))

/usr/local/lib/python3.10/dist-packages/networkx/classes/coreviews.py in __getitem__(self, key)
     51 
     52     def __getitem__(self, key):
---> 53         return self._atlas[key]
     54 
     55     def copy(self):

KeyError: 'label'
```
</details>

The problem is that https://github.com/networkx/networkx/blob/82efb9e69d5ab72d75de53afe134c6cda13b18ff/networkx/algorithms/graph_hashing.py#L35 accesses the edge attributes using `G[node][nbr][edge_attr]` which is not valid for `MultiDiGraph`s where there is more than one edge.

### Expected Behavior

**Two options:**
1. Detect if the input graph is a` MultiDiGraph` and `edge_attr is not None` and raise a proper error message such as `'edge_attr' argument is not supported for MultiDiGraph`
2. Implement the special case in `_neighborhood_aggregate` to properly support MultiDiGraphs.

In addition, unit tests should be added to make sure the function is working as expected for these edge cases.

I am happy to submit a patch for this bug. Please let me know if option (1) or (2) would be preferable.

### Steps to Reproduce

**This works as expected:**

```python
G1 = nx.Graph()

G1.add_edges_from(
    [
        (1, 2, {"label": "A"}),
        (2, 3, {"label": "A"}),
        (3, 1, {"label": "A"}),
        (1, 4, {"label": "B"}),
    ]
)
nx.weisfeiler_lehman_graph_hash(G1)  # ok
nx.weisfeiler_lehman_graph_hash(G1, edge_attr="label")  # ok
```

**This fails (note that the `nx.Graph` was replaced with `nx.MultiDiGraph`):**

```python
G1 = nx.MultiDiGraph()

G1.add_edges_from(
    [
        (1, 2, {"label": "A"}),
        (2, 3, {"label": "A"}),
        (3, 1, {"label": "A"}),
        (1, 4, {"label": "B"}),
    ]
)

nx.weisfeiler_lehman_graph_hash(G1)  # ok
nx.weisfeiler_lehman_graph_hash(G1, edge_attr="label")  # error
```

### Environment

Python version: `Python 3.10.12`
NetworkX version: `networkx==3.3`

### Additional context

N/A

**Comment by dschult** [comment]
Thanks for submitting this issue. I agree that this is not a good error message for MultiDiGraph input. And I'm not familiar enough with the algorithm to know whether MultiGraphs should even be considered using this technique. Of course, we can always construct a way to handle multiedges. But will it help make a reasonable hash?

Do you know of any articles which build off the papers cited in the doc_string (or not) that handle hashing of multigraphs?

The easy fix (that can be expanded later) would be to check `G.is_multigraph()` early on and raise an exception. There is a decorator to make this simpler:  `@nx.utils.not_implemented_for("multigraph")`

**Comment by PhilippvK** [comment]
@dschult I had the same thought. Fixing the edge_attr code for multigraphs is trivial, but I do not know if the algorithm/code is supposed to yield correct results for this kind of graph.

We can, of course, do some quantitative experiments to see if the resulting hashes are as expected, but this would not prove that there are no counterexamples.

For now, I can just implement the fix by adding the mentioned decorator. Thanks for the tip!

**Comment by Peiffap** [comment]
I believe this was fixed in #7614 and can be closed now.
